File: Data Structures & Algorithms/mergeSort/submission-1.py
# Definition for a pair.
# class Pair:
#     def __init__(self, key: int, value: str):
#         self.key = key
#         self.value = value
class Solution:
    def merge(self, pairs: List[Pair], s, m, e):
        L = pairs[s : m + 1]
        R = pairs[m + 1 : e + 1]
        i, j, k = 0, 0, s
        l = len(L)
        r = len(R)
        while(i < l and j < r):
            if(L[i].key <= R[j].key):
                pairs[k] = L[i]
                i += 1
                k += 1
            else:
                pairs[k] = R[j]
                j += 1
                k += 1
        while(i < l):
            pairs[k] = L[i]
            i += 1
            k += 1
        while(j < r):
            pairs[k] = R[j]
            j += 1
            k += 1

    def mergeSortHelper(self, pairs: List[Pair], s, e) -> List[Pair]:
        if(e - s + 1) <= 1:
            return pairs
    
        m = (e + s) // 2
      
        self.mergeSortHelper(pairs, s, m)
        self.mergeSortHelper(pairs, m + 1, e)
        self.merge(pairs, s, m, e)

        return pairs

    def mergeSort(self, pairs: List[Pair]) -> List[Pair]:
        return self.mergeSortHelper(pairs, 0, len(pairs) - 1)

File: Data Structures & Algorithms/quickSort/submission-1.java
// Definition for a pair.
// class Pair {
//     int key;
//     String value;
//
//     public Pair(int key, String value) {
//         this.key = key;
//         this.value = value;
//     }
// }
class Solution {
    public List<Pair> quickSort(List<Pair> pairs) {
        quickSortHelper(pairs, 0, pairs.size() - 1);
        return pairs;
    }

    private void quickSortHelper(List<Pair> pairs, int s, int e) {
        if(s >= e) {
            return;
        }

        int pivot = e;
        int i = s;
        Pair temp = null;

        for(int j = s; j < e; j++) {
            if(pairs.get(j).key < pairs.get(pivot).key) {
                // temp = pairs[i];
                // pairs[i] = pairs[j];
                // pairs[j] = temp;
            
                swap(pairs, i, j);
                i++;
            }
        }

        // temp = pairs.get(i);
        // pairs.set(i, pairs.get(pivot));
        // pairs.set(pivot, temp);
        swap(pairs, i, pivot);

        quickSortHelper(pairs, s, i - 1);
        quickSortHelper(pairs, i + 1, e);
    }

    private void swap(List<Pair> pairs, int index1, int index2) {
        Pair temp = null;
        temp = pairs.get(index1);
        pairs.set(index1, pairs.get(index2));
        pairs.set(index2, temp);
    }
}
